From 7181cc8cb6f6f0685fee11edfd84ca4d453fcbf6 Mon Sep 17 00:00:00 2001
From: Aihua Xu <aihuaxu@apache.org>
Date: Wed, 27 Jul 2016 16:08:34 -0400
Subject: [PATCH 664/723] CDH-42673 HIVE-14296: Session count is not
 decremented when HS2 clients do not shutdown
 cleanly. (Naveen Gangam, reviewed by Szehon Ho and
 Mohit Sabharwal)

Conflicts:
	service/src/java/org/apache/hive/service/cli/session/SessionManager.java

Change-Id: Ic863395e1f311a1e353cb799dc7d52ac136d2454
---
 .../hive/service/cli/session/SessionManager.java   |    4 +++-
 .../hive/service/cli/thrift/ThriftCLIService.java  |    3 ---
 2 files changed, 3 insertions(+), 4 deletions(-)

diff --git a/service/src/java/org/apache/hive/service/cli/session/SessionManager.java b/service/src/java/org/apache/hive/service/cli/session/SessionManager.java
index 79b39ef..68adb34 100644
--- a/service/src/java/org/apache/hive/service/cli/session/SessionManager.java
+++ b/service/src/java/org/apache/hive/service/cli/session/SessionManager.java
@@ -308,14 +308,16 @@ public SessionHandle openSession(TProtocolVersion protocol, String username, Str
       throw new HiveSQLException("Failed to execute session hooks", e);
     }
     handleToSession.put(session.getSessionHandle(), session);
+    LOG.info("Session opened, " + session.getSessionHandle() + ", current sessions:" + getOpenSessionCount());
     return session.getSessionHandle();
   }
 
-  public void closeSession(SessionHandle sessionHandle) throws HiveSQLException {
+  public synchronized void closeSession(SessionHandle sessionHandle) throws HiveSQLException {
     HiveSession session = handleToSession.remove(sessionHandle);
     if (session == null) {
       throw new HiveSQLException("Session does not exist: " + sessionHandle);
     }
+    LOG.info("Session closed, " + sessionHandle + ", current sessions:" + getOpenSessionCount());
     try {
       session.close();
     } finally {
diff --git a/service/src/java/org/apache/hive/service/cli/thrift/ThriftCLIService.java b/service/src/java/org/apache/hive/service/cli/thrift/ThriftCLIService.java
index eb0650a..8bb9c1f 100644
--- a/service/src/java/org/apache/hive/service/cli/thrift/ThriftCLIService.java
+++ b/service/src/java/org/apache/hive/service/cli/thrift/ThriftCLIService.java
@@ -72,7 +72,6 @@
   protected CLIService cliService;
   private static final TStatus OK_STATUS = new TStatus(TStatusCode.SUCCESS_STATUS);
   protected static HiveAuthFactory hiveAuthFactory;
-  private static final AtomicInteger sessionCount = new AtomicInteger();
 
   protected int portNum;
   protected InetAddress serverIPAddress;
@@ -324,7 +323,6 @@ public TOpenSessionResp OpenSession(TOpenSessionReq req) throws TException {
       if (context != null) {
         context.setSessionHandle(sessionHandle);
       }
-      LOG.info("Opened a session " + sessionHandle + ", current sessions: " + sessionCount.incrementAndGet());
     } catch (Exception e) {
       LOG.warn("Error opening session: ", e);
       resp.setStatus(HiveSQLException.toTStatus(e));
@@ -462,7 +460,6 @@ public TCloseSessionResp CloseSession(TCloseSessionReq req) throws TException {
     try {
       SessionHandle sessionHandle = new SessionHandle(req.getSessionHandle());
       cliService.closeSession(sessionHandle);
-      LOG.info("Closed a session " + sessionHandle + ", current sessions: " + sessionCount.decrementAndGet());
       resp.setStatus(OK_STATUS);
       ThriftCLIServerContext context =
         (ThriftCLIServerContext)currentServerContext.get();
-- 
1.7.9.5

